[STEP 02] 박스오피스 앱 ― D\.O\., Jenna

ueunli opened the pull request:
안녕하세요 Cory!! @corykim0829
`디오` & `제나` 입니다👋

`STEP2`에서는 일일 박스오피스와 개별 영화 API를 요청하는 `네트워크 모델을 구현하는 단계`였습니다. 저희는 `url`과 `urlRequest를` 생성할 수 있는 타입과 `dataTask`를 `request`하는 타입으로 나누어 만들어 보았습니다.

추가로 STEP2-1(도전스텝)에서 테스트와 관련된 내용이 있었으나, `Network layer`에 대한 추가 학습 이후 도전해보기로 했어요. 잘 부탁드립니다 🙇🏻‍♀️🙇🏻‍♂️!!

---

　
# 요구 사항 및 구현

### 요구 사항
- [x] 네트워크 통신을 담당할 타입을 설계하고 구현합니다.
- [x] 영화진흥위원회의 일별 박스오피스 API 문서의 데이터 형식을 고려하여 서버와 실제로 데이터를 주고받도록 구현합니다.
   - [x] 오늘의 일일 박스오피스 조회
   - [x] 영화 개별 상세 조회

### 구현
<img src="https://i.imgur.com/ZfLTArR.png" width="40%"></img>
- **Network**
    - `API`:  네트워크 통신에 필요한 엔드포인트, `url`, `urlRequest` 생성 
    - `NetworkError`: 네트워크시 발생한 에러타입 정의, `LocalizedError` 채택
    - `HTTPMethodType`: `get`, `post`, `put`, `delete` 정의
    - `NetworkServable`: `dataTask` 메서드를 정의한 프로토콜 구현
- **Model**
    - `DailyBoxOfficeDTO`: 일일박스오피스 정의한 DTO
    - `MovieDetailDTO`: 개별영화 데이터를 정의한 DTO
- **View**
    - `LaunchScreen`
    - `Main`
- **Controllers**
    - `ViewController`


　
# 고민했던 부분
- **ViewController의 저장속성**
   - 추후 ViewController가 DataSource로서 기능하기 용이하도록 보다 정제된 형태의 DTO를 가지게 하는 편이 맞다고 생각했으나, 
   - 현재 요구사항이 '데이터를 주고받도록 구현'임을 고려하여 추가적인 네트워크 레이어를 구현하지 않았으므로, 뷰컨트롤러가 임시로 `NetworkServable`을 채택, 현 DTO들을 갖도록 구현했어요
   - → 다음 스텝에서 저장속성 타입 등을 수정할 계획입니다!
- **Endpoint**
   - 지금 구현에서는 `API` 열거형이 Endpoint의 역할을 겸하고 있는데요, 
   - `API`의 로직이 다소 무거워진 점이 걸려 리팩토링을 하려고 했으나, 앞으로 추가될 api 조건도 고려해야 될 것 같아 다음 스텝 혹은 Test Double 적용 과정에서 진행하는 게 좋겠다고 판단했습니다..!


　
　

altokeys committed:
Chore: 파일명 변경(+DTO)

altokeys committed:
Feat: NetworkConfigure 구현
- NetworkConfigurable 프로토콜 정의

altokeys committed:
Feat: EndPoint 구현
- NetworkRequestable 프로토콜 정의
- NetworkError 정의

altokeys committed:
Refactor: endpoint 구현 로직 변경

altokeys committed:
Feat: 일별 박스오피스 API 호출 기능 구현

ueunli committed:
Feat: 영화 상세 API 호출 로직 구현

ueunli committed:
Refactor: api 요청 메서드 로직 수정

- 컴플리션 핸들러의 매개변수를 Result타입으로 변경
- NetworkError에 LocalizedError프로토콜 채택

ueunli committed:
Chore: ViewController가 MovieDetailDTO를 가지도록 프로퍼티 추가

- movie 프로퍼티 추가
- movie, boxOffice 프로퍼티를 private화

altokeys committed:
Chore: DTO 폴더 분리

altokeys committed:
Chore: 네이밍 변경
- boxOffice -> dailyBoxOffice
- movie -> moveDetail

ueunli committed:
Refactor: NetworkService구조체를 NetworkServable프로토콜로 변경

- api로 데이터를 요청하고 디코딩하는 두 메서드를 하나로 통합하여 중복로직 최소화
- ViewController에서 해당 프로토콜을 채택

ueunli committed:
Chore: NetworkRepresentable 프로토콜 파일 분리

corykim0829 commented on BoxOffice/Controllers/ViewController.swift:
```diff
@@ -7,13 +7,27 @@
+        fetchData(apiType: api) { result in
+            switch result {
+            case .success(let data): self.dailyBoxOffice = data as? DailyBoxOfficeDTO
+            case .failure(let error): print(error.localizedDescription)
```
case의 `:` 뒤에는 개행을 해주어서 확실하게 알아볼 수 있도록 하면 좋을 것 같아요!

ueunli replied:
수정하였습니다 🙆🏻‍♀️

> 해당 커밋: 7da679b

corykim0829 commented on BoxOffice/Network/NetworkServable.swift:
```diff
@@ -0,0 +1,63 @@
+
+protocol NetworkServable {
+    
+    func fetchData(apiType: API, completion: @escaping (Result<APIRepresentable, NetworkError>) -> Void)
```
NetworkServable을 프로토콜로 가져가는 이유가 어떻게 되나요? 프로토콜로 사용하면 어떤 장점이 있나요?

completion 클로저에서 APIRepresentable 프로토콜 타입으로 반환하는 부분은 사용할 때마다 타입 변환을 한번 더 옵셔널 처리해야한다는 번거로움이 있는 것 같아요.

원하는 타입으로 반환하는 방법을 고민해보면 좋을 것 같아요.

corykim0829 commented on BoxOffice/Network/API.swift:
```diff
@@ -0,0 +1,79 @@
+
+import Foundation
+
+enum API {
```
이렇게 enum으로 구현하면, 나중에 요청할 케이스가 많아지면 유지보수가 힘들어지게 될 것 같네요
하나의 케이스를 추가하거나 제거하게 되면, 모든 switch를 반영해줘야해서 불편함이 있으니 참고하세요!

enum + case으로 사용하는 것은 확실한 케이스가 나뉘어지는 변수처럼 사용하는게 좋습니다!

ueunli replied:
의논 끝에, 케이스가 많아지면 말씀해주신 문제점이 두드러질 것 같아 프로토콜 및 case별 구조체로 리팩토링 했습니다!

> 해당 커밋: 373bef5

corykim0829 commented on BoxOffice/Network/NetworkError.swift:
```diff
@@ -0,0 +1,32 @@
+
+extension NetworkError: LocalizedError {
+
+    var errorDescription: String? {
```
다양한 에러 케이스를 고려해서 description까지 처리한 부분이 좋네요 👍

corykim0829 reviewed (changes requested):
수고하셨습니다!!

다음 스텝에 해야할 게 많아보이지만 같이 이야기해보고 넘어가면 좋을 것 같아요!

커멘트 확인해주세요!

altokeys committed:
Chore: 코드 개행

ueunli committed:
Refactor: API열거형을 프로토콜로 변경 및 제네릭 메서드를 통해 OCP 보완

- 폴더링
- API열거형 → 연관타입을 갖는 EndPointProtocol 및 이를 준수하는 api별 구조체
- NetworkServable 프로토콜에 제네릭 메서드 fetchData(endPoint:completion:) 구현

corykim0829 commented on BoxOffice/Controllers/ViewController.swift:
```diff
@@ -8,12 +8,32 @@
-
-
+    
+    func updateData<T: APIRepresentable, E: EndPointProtocol>(from api: E, completion: @escaping (T?) -> Void) where E.DTO == T {
```
EndPointProtocol에 APIRepresentable 타입이 있어서 해당 타입을 굳이 제너릭 선언을 할 필요가 없어보여요

corykim0829 reviewed (approved):
수고하셨습니다!!!

다음 스텝 진행 해주시고

남겨드린 커멘트 반영해주시고, 커멘트에 대한 의견 있으시면 답글로 남겨주세요!
